### PR TITLE
Add tags to the emr jobs created based on the query.

### PR DESCRIPTION
**Pull request by vamsi-amazon**
### Description
* Added support to extract indexName, tableName and datasourceName in Async Queries.
* Added tags to emr jobs submitted and changed jobName as per the query.
* Added datasource Authorization.
* Removed hardcoded my_glue datasource from earlier queries.
* Throwing unsupported exceptions in case of queries with no datasource and also for Async Requests with PPL Language and other datasource connectors.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 1]
```diff
@@ -106,4 +129,80 @@ private String constructSparkParameters(String datasourceName) throws URISyntaxE
...
+    StartJobRequest startJobRequest;
+    FullyQualifiedTableName fullyQualifiedTableName =
+        SQLQueryUtils.extractFullyQualifiedTableName(dispatchQueryRequest.getQuery());
+    if (fullyQualifiedTableName == null || fullyQualifiedTableName.getDatasourceName() == null) {
+      throw new UnsupportedOperationException("Queries without a datasource are not supported");
+    }
+    dataSourceUserAuthorizationHelper.authorizeDataSource(
+        this.dataSourceService.getRawDataSourceMetadata(
+            fullyQualifiedTableName.getDatasourceName()));
+    String jobName =
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:144]: Query type will decide job name? We could refactor the code later.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 2]
```diff
@@ -106,4 +129,80 @@ private String constructSparkParameters(String datasourceName) throws URISyntaxE
...
+            dispatchQueryRequest.getQuery(),
+            jobName,
+            dispatchQueryRequest.getApplicationId(),
+            dispatchQueryRequest.getExecutionRoleARN(),
+            constructSparkParameters(fullyQualifiedTableName.getDatasourceName()),
+            tags);
+    return startJobRequest;
+  }
+
+  private StartJobRequest getStartJobRequestForIndexRequest(
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:161]: Only Create Index is handled in this function, and submit SQL to EMR-S. But DROP INDX, DESCRIBE INDEX will not. right?

Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:161]: DROP INDEX will be handled in next PR. Will refactor this class then. 
DESCRIBE INDEX is also job submission right? @penghuo

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 1 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:144]: Yes.

File: spark/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import org.opensearch.sql.spark.antlr.parser.SqlBaseParser;
+import org.opensearch.sql.spark.antlr.parser.SqlBaseParserBaseVisitor;
+import org.opensearch.sql.spark.dispatcher.model.FullyQualifiedTableName;
+import org.opensearch.sql.spark.dispatcher.model.IndexDetails;
+
+@UtilityClass
+public class SQLQueryUtils {
+
+  public static FullyQualifiedTableName extractFullyQualifiedTableName(String sqlQuery) {
+    return trySparkSQL(sqlQuery);
```
[the remark this answers is not in the document]
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java]: Good one. No answer right now. The second one will be picked up right now.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 4]
```diff
@@ -106,4 +129,80 @@ private String constructSparkParameters(String datasourceName) throws URISyntaxE
         "spark.sql.catalog." + datasourceName, FLINT_DELEGATE_CATALOG);
     return s3GlueSparkSubmitParameters.toString();
   }
+
+  private StartJobRequest getStartJobRequestForNonIndexQueries(
+      DispatchQueryRequest dispatchQueryRequest) {
+    StartJobRequest startJobRequest;
+    FullyQualifiedTableName fullyQualifiedTableName =
+        SQLQueryUtils.extractFullyQualifiedTableName(dispatchQueryRequest.getQuery());
+    if (fullyQualifiedTableName.getDatasourceName() == null) {
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:138]: Do we need to add check for query with multiple table name present? I assume we don't support it for now?

File: spark/build.gradle [hunk 5]
```diff
@@ -7,13 +7,42 @@ plugins {
...
 repositories {
     mavenCentral()
 }
 
+tasks.register('downloadG4Files', Exec) {
+    description = 'Download remote .g4 files from GitHub'
+
+    executable 'curl'
+
+//    Need to add these back once the grammar issues with indexName and tableName is addressed in flint integration jar.
```
Comment (dai-chen) [inline spark/build.gradle:22]: I'm working on this. Will let you know once PR merged.

**Comment by dai-chen** [comment]
Could you add missing Javadoc for new main interface?

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 4 above)
Comment (vamsi-amazon) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:138]: yes, I am assuming a single table name. Need to handle it in a separate PR.
Will make a TODO comment in SQLQueryUtils.

**Comment by vamsi-amazon** [comment]
> Could you add missing Javadoc for new main interface?

Addressed.

**Review by penghuo — APPROVED** [review]

**Review by kavithacm — APPROVED** [review]

**Pull request merged**